qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/HandlerThreadActivity.kt

```diff
@@ -0,0 +1,561 @@
+    private lateinit var binding: LayoutActivityHandlerThreadBinding
+
+    /* ----- Mapbox Maps components ----- */
+    private lateinit var mapboxMap: MapboxMap
+
+    /* ----- Mapbox Navigation components ----- */
+    private lateinit var mapboxNavigation: MapboxNavigation
+
+    // location puck integration
+    private val navigationLocationProvider = NavigationLocationProvider()
+
+    // camera
+    private lateinit var navigationCamera: NavigationCamera
+    private lateinit var viewportDataSource: MapboxNavigationViewportDataSource
+    private val pixelDensity = Resources.getSystem().displayMetrics.density
+    private val overviewPadding: EdgeInsets by lazy {
+        EdgeInsets(
+            140.0 * pixelDensity,
+            40.0 * pixelDensity,
+            120.0 * pixelDensity,
+            40.0 * pixelDensity
+        )
+    }
+    private val landscapeOverviewPadding: EdgeInsets by lazy {
+        EdgeInsets(
+            30.0 * pixelDensity,
+            380.0 * pixelDensity,
+            20.0 * pixelDensity,
+            20.0 * pixelDensity
+        )
+    }
+    private val followingPadding: EdgeInsets by lazy {
+        EdgeInsets(
+            180.0 * pixelDensity,
+            40.0 * pixelDensity,
+            150.0 * pixelDensity,
+            40.0 * pixelDensity
+        )
+    }
+    private val landscapeFollowingPadding: EdgeInsets by lazy {
+        EdgeInsets(
+            30.0 * pixelDensity,
+            380.0 * pixelDensity,
+            110.0 * pixelDensity,
+            40.0 * pixelDensity
+        )
+    }
+
+    // trip progress bottom view
+    private lateinit var tripProgressApi: MapboxTripProgressApi
+
+    // voice instructions
+    private var isVoiceInstructionsMuted = false
+    private lateinit var maneuverApi: MapboxManeuverApi
+    private lateinit var speechAPI: MapboxSpeechApi
+    private lateinit var voiceInstructionsPlayer: MapboxVoiceInstructionsPlayer
+
+    // route line
+    private lateinit var routeLineAPI: MapboxRouteLineApi
+    private lateinit var routeLineView: MapboxRouteLineView
+    private lateinit var routeArrowView: MapboxRouteArrowView
+    private val routeArrowAPI: MapboxRouteArrowApi = MapboxRouteArrowApi()
+
+    /* ----- Voice instruction callbacks ----- */
+    private val voiceInstructionsObserver =
+        VoiceInstructionsObserver { voiceInstructions ->
+            runOnUiThread {
+                speechAPI.generate(
+                    voiceInstructions,
+                    speechCallback
+                )
+            }
+        }
+
+    private val voiceInstructionsPlayerCallback =
+        MapboxNavigationConsumer<SpeechAnnouncement> { value ->
+            runOnUiThread {
+                // remove already consumed file to free-up space
+                speechAPI.clean(value)
+            }
+        }
+
+    private val speechCallback =
+        MapboxNavigationConsumer<Expected<SpeechError, SpeechValue>> { expected ->
+            runOnUiThread {
```

Comment:
BTW, do the customers understand that they will have to invoke `runOnUiThread` once in a while?
I think they didn't like the first approach because they had to think about threads on their side. But they will still have to now, although less.